client/llb/marshal.go

```diff
@@ -3,38 +3,70 @@ package llb
 import (
 	"encoding/binary"
 	"io"
+
+	"github.com/moby/buildkit/solver/pb"
+	digest "github.com/opencontainers/go-digest"
 )
 
-// These are temporary functions for testing the solver without frontends
+// Definition is the LLB definition structure with per-vertex metadata entries
+// Corresponds to the Definition structure defined in solver/pb.Definition.
+type Definition struct {
+	Def      [][]byte
+	Metadata map[digest.Digest]OpMetadata
+}
+
+func (def Definition) ToPB() *pb.Definition {
+	md := make(map[digest.Digest]pb.OpMetadata)
+	for k, v := range def.Metadata {
+		md[k] = v.OpMetadata
+	}
+	return &pb.Definition{
+		Def:      def.Def,
+		Metadata: md,
+	}
+}
+
+func (def *Definition) FromPB(x *pb.Definition) {
+	def.Def = x.Def
+	def.Metadata = make(map[digest.Digest]OpMetadata)
+	for k, v := range x.Metadata {
+		def.Metadata[k] = OpMetadata{v}
+	}
+}
+
+type OpMetadata struct {
+	pb.OpMetadata
+}
 
-func WriteTo(dt [][]byte, w io.Writer) error {
-	for _, d := range dt {
-		l := make([]byte, 4)
-		binary.LittleEndian.PutUint32(l, uint32(len(d)))
-		if _, err := w.Write(l); err != nil {
-			return err
-		}
-		if _, err := w.Write(d); err != nil {
-			return err
-		}
+func WriteTo(def Definition, w io.Writer) error {
+	b, err := def.ToPB().Marshal()
+	if err != nil {
+		return err
 	}
-	return nil
+	l := make([]byte, 4)
```

Comment:
There is probably no reason to add a length if we just marshal the whole object. Just write/read the protobuf object.